presto-main/src/main/java/com/facebook/presto/memory/HighMemoryTaskKiller.java

```diff
@@ -0,0 +1,248 @@
+import io.airlift.units.Duration;
+import org.jheaps.annotations.VisibleForTesting;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.management.JMException;
+import javax.management.Notification;
+import javax.management.NotificationListener;
+import javax.management.ObjectName;
+import javax.management.openmbean.CompositeData;
+
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryUsage;
+import java.util.AbstractMap;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import static com.facebook.presto.memory.HighMemoryTaskKillerStrategy.FREE_MEMORY_ON_FREQUENT_FULL_GC;
+import static com.facebook.presto.memory.HighMemoryTaskKillerStrategy.FREE_MEMORY_ON_FULL_GC;
+import static com.facebook.presto.spi.StandardErrorCode.EXCEEDED_HEAP_MEMORY_LIMIT;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableListMultimap.toImmutableListMultimap;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HighMemoryTaskKiller
+{
+    private static final Logger log = Logger.get(HighMemoryTaskKiller.class);
+    private static final String GC_NOTIFICATION_TYPE = "com.sun.management.gc.notification";
+    private static final MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
+    private final NotificationListener gcNotificationListener = (notification, ignored) -> onGCNotification(notification);
+    private final SqlTaskManager sqlTaskManager;
+    private final HighMemoryTaskKillerStrategy taskKillerStrategy;
+    private final boolean taskKillerEnabled;
+    private final Duration taskKillerFrequentFullGCDurationThreshold;
+    private long lastFullGCTimestamp;
+    private long lastFullGCCollectedBytes;
+    private final long reclaimMemoryThreshold;
+    private final long heapMemoryThreshold;
+    Ticker ticker;
+
+    @Inject
+    public HighMemoryTaskKiller(SqlTaskManager sqlTaskManager, TaskManagerConfig taskManagerConfig)
+    {
+        requireNonNull(taskManagerConfig, "taskManagerConfig is null");
+
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager must not be null");
+
+        this.taskKillerStrategy = taskManagerConfig.getHighMemoryTaskKillerStrategy();
+        this.taskKillerEnabled = taskManagerConfig.isHighMemoryTaskKillerEnabled();
+
+        this.taskKillerFrequentFullGCDurationThreshold = taskManagerConfig.getHighMemoryTaskKillerFrequentFullGCDurationThreshold();
+        this.reclaimMemoryThreshold = (long) (memoryMXBean.getHeapMemoryUsage().getMax() * taskManagerConfig.getHighMemoryTaskKillerGCReclaimMemoryThreshold());
```

Comment:
I am not sure about this: Do we want to reclaim 1% (as stated above) of the Xmx or the current heap usage ? The Xmx can be arbitrary high ? I believe we should threshhold on reclaiming say at least 1% of the current heap usage.

Reply:
Added detail in the config that explains that this is based on -Xmx.